### PR TITLE
Fix for exponential time decaying window functions

### PR DESCRIPTION
**Pull request by excitoon**
### Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix for exponential time decaying window functions. Now respecting boundaries of the window.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by novikd — APPROVED** [review]

File: src/Processors/Transforms/WindowTransform.cpp [hunk 1]
```diff
@@ -1607,82 +1563,78 @@ namespace recurrent_detail
...
+{
+    RowNumber previous_frame_start;
+    RowNumber previous_frame_end;
+    Float64 previous_time;
+    Float64 previous_sum;
+    Float64 previous_count;
 };
 
-struct WindowFunctionExponentialTimeDecayedSum final : public RecurrentWindowFunction<1>
+struct WindowFunctionExponentialTimeDecayedSum final : public StatefulWindowFunction<ExponentialTimeDecayedSumState>, public WindowFunctionHelpers
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Processors/Transforms/WindowTransform.cpp:1630]: Yes, it's not needed.

File: src/Processors/Transforms/WindowTransform.cpp [hunk 2]
```diff
@@ -1607,82 +1563,78 @@ namespace recurrent_detail
...
-        for (size_t i = 0; i < state_size; ++i)
-            state[i] = 0.0;
-    }
-
-    template<typename T>
-    static T getLastValueFromInputColumn(const WindowTransform * transform, size_t function_index, size_t column_index)
-    {
-        return recurrent_detail::getLastValueFromInputColumn<T>(transform, function_index, column_index);
+        auto * const state = static_cast<State *>(static_cast<void *>(place));
+        *state = State();
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Processors/Transforms/WindowTransform.cpp:1598]: Cool!

File: src/Processors/Transforms/WindowTransform.cpp [hunk 3]
```diff
@@ -1792,32 +1773,35 @@ struct WindowFunctionExponentialTimeDecayedMax final : public RecurrentWindowFun
     void windowInsertResultInto(const WindowTransform * transform,
         size_t function_index) override
     {
-        Float64 last_max = getLastValueFromState<Float64>(transform, function_index, STATE_MAX);
-        Float64 last_t = getLastValueFromInputColumn<Float64>(transform, function_index, ARGUMENT_TIME);
+        Float64 result = -std::numeric_limits<Float64>::infinity();
```
Comment (novikd) [inline src/Processors/Transforms/WindowTransform.cpp:1776]:
```suggestion
        Float64 result = std::numeric_limits<Float64>::lowest();
```

**Review by novikd — APPROVED** [review]

File: src/Processors/Transforms/WindowTransform.cpp (hunk 3 above)
Comment (excitoon) [reply src/Processors/Transforms/WindowTransform.cpp]: Why is it better?

**Pull request merged**